jvb/src/main/kotlin/org/jitsi/videobridge/cc/av1/Av1DDAdaptiveSourceProjectionContext.kt

```diff
@@ -0,0 +1,648 @@
+                seqGap += seqGap(f1, f2)
+                f1 = f2
+            } while (f2 !== frame)
+            /* refFrame is a projected frame, so it has a projection. */
+            refSeq = refFrame.projection!!.latestProjectedSeqNum
+        } else {
+            /* refFrame is later than frame in decode order. */
+            do {
+                f2 = prevFrame(f1)
+                checkNotNull(f2) {
+                    "No previous frame found before frame with frame number ${f1.frameNumber}, " +
+                        "even though refFrame ${refFrame.frameNumber} is after " +
+                        "frame ${frame.frameNumber}!"
+                }
+                seqGap += -seqGap(f2, f1)
+                f1 = f2
+            } while (f2 !== frame)
+            refSeq = refFrame.projection!!.earliestProjectedSeqNum
+        }
+
+        val projectedSeq = RtpUtils.applySequenceNumberDelta(refSeq, seqGap)
+        val projectedTs = RtpUtils.applyTimestampDelta(refFrame.projection!!.timestamp, tsGap)
+        val projectedFrameNumber = RtpUtils.applySequenceNumberDelta(refFrame.projection!!.frameNumber, frameNumGap)
+
+        return Av1DDFrameProjection(
+            diagnosticContext = diagnosticContext,
+            av1Frame = frame,
+            ssrc = lastAv1FrameProjection.ssrc,
+            timestamp = projectedTs,
+            sequenceNumberDelta = RtpUtils.getSequenceNumberDelta(projectedSeq, initialPacket.sequenceNumber),
+            frameNumber = projectedFrameNumber,
+            templateIdDelta = lastAv1FrameProjection.templateIdDelta,
+            dti = newDt?.let { frame.structure?.getDtBitmaskForDt(it) },
+            mark = mark,
+            created = receivedTime
+        )
+    }
+
+    override fun needsKeyframe(): Boolean {
+        if (av1QualityFilter.needsKeyframe) {
+            return true
+        }
+
+        return lastAv1FrameProjection.av1Frame == null
+    }
+
+    override fun rewriteRtp(packetInfo: PacketInfo) {
+        if (packetInfo.packet !is Av1DDPacket) {
+            logger.info("Got a non-AV1 DD packet.")
+            throw RewriteException("Non-AV1 DD packet in AV1 DD source projection")
+        }
+        val av1Packet = packetInfo.packetAs<Av1DDPacket>()
+
+        val av1Frame = lookupAv1Frame(av1Packet)
+            ?: throw RewriteException("Frame not in tracker (aged off?)")
+
+        val av1Projection = av1Frame.projection
+            ?: throw RewriteException("Frame does not have projection?")
+        /* Shouldn't happen for an accepted packet whose frame is still known? */
+
+        av1Projection.rewriteRtp(av1Packet)
+    }
+
+    override fun rewriteRtcp(rtcpSrPacket: RtcpSrPacket): Boolean {
+        val lastAv1FrameProjectionCopy: Av1DDFrameProjection = lastAv1FrameProjection
+        if (rtcpSrPacket.senderSsrc != lastAv1FrameProjectionCopy.av1Frame?.ssrc) {
+            return false
+        }
+
+        rtcpSrPacket.senderSsrc = lastAv1FrameProjectionCopy.ssrc
```

Comment:
Given the check above this only does anything if `lastAv1FrameProjectionCopy.av1Frame.ssrc != lastAv1FrameProjectionCopy.ssrc`. Is that expected?

Reply:
Yes - the idea here is that for multi-encoding simulcast all (three) of the incoming SSRCs are mapped to a single outgoing SSRC.